Better composable XMLs support

I'd like to compose XMLs with functions. That works but solution is a bit noisy:

```rust
use format_xml::xml;

fn f(x: u8, y: u8) -> String {
    let result = xml! {
        <A>{x}</A>
        <B>{y}</B>
    }
    .to_string();
    result
}

fn main() {
    let x = xml! {
        <C>
            {f(1, 2)}
        </C>
        <C>
            {f(2, 1)}
        </C>
    }
    .to_string();
    println!("{}", x);
}
```

Removing intermediate `result` variable
```rust
fn f(x: u8, y: u8) -> String {
    xml! {
        <A>{x}</A>
        <B>{y}</B>
    }
    .to_string()
}
```
causes unobvious compiler error:
```
error: expected expression, found `.`
 --> src/main.rs:8:5
  |
8 |     .to_string()
  |     ^ expected expression

error[E0308]: mismatched types
 --> src/main.rs:4:5
  |
4 | /     xml! {
5 | |         <A>{x}</A>
6 | |         <B>{y}</B>
7 | |     }
  | |     ^
  | |     |
  | |_____expected `()`, found struct `std::fmt::Arguments`
  |       help: consider using a semicolon here
  |
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

Which can be fixed by wrapping `xml!` macros usage with `()` or `{}`:
```rust
fn f(x: u8, y: u8) -> String {
    (xml! {
        <A>{x}</A>
        <B>{y}</B>
    })
    .to_string()
}
```

Returning `std::fmt::Arguments` without constructing intermediate `String` would be even better, but I didn't find a way to satisfy borrow checker.

Thanks for opening this issue!

The unobvious compile error is mysterious. The macro expands to

```rust
format_args !
          (concat!(concat!(concat!(concat!(concat!(concat!(concat!(concat!(concat!(concat!(concat!(concat!(concat!(concat!("", "<"), stringify ! (A)), ">"), "{}"),
                  "</"),
                  stringify ! (A)),
                  ">"),
                  "<"),
                  stringify ! (B)),
                  ">"),
                  "{}"),
                  "</"),
                  stringify ! (B)),
                  ">"),
           x, y)
```

When I manually replace the `xml!` invocation with that monstrosity then everything works as expected! I suspect a bug in Rust's parser interaction with macro output but I will need to do a deeper investigation,

In any case there's a better way to implement this. As you've noticed you _cannot_ return `std::fmt::Arguments` by value as they capture their arguments by reference. These are references to local variables in your function and won't be valid when your function returns.

For the specific purpose of composability I added the following syntax:

```rust
// Can also accept &mut std::fmt::Formatter directly
fn compose(f: &mut dyn std::fmt::Write, x: u8, y: u8) -> std::fmt::Result {
    f.write_fmt(xml! {
        <A>{x}</A>
        <B>{y}</B>
    })
}

fn main() {
    xml::println! {
        <C>
            |f| {compose(f, 1, 2)}
        </C>
        <C>
            |f| {compose(f, 1, 2)}
        </C>
    };
}
```

The idea is that you can pass the underlying formatter through just like if you would write an fmt trait implementation. This should be sufficient for composability. So instead of returning an Arguments structure, accept a formatter and write directly into it.

That said when I run this code I get `error[E0658]: attributes on expressions are experimental` because I try to wrap the resulting println's in a `#[inline(never)]` annotation to keep all the code generated for the formatting contained. No idea why I didn't catch this earlier. Using the `xml! { .. }.to_string()` does not have this problem.

--- a/readme.md
+++ b/readme.md
@@ -7,11 +7,11 @@ Template XML formatting
 
 Minimal compiletime templating for XML in Rust!
 
-The `xml!` macro accepts an XML-like syntax and transforms it into a `format_args!` invocation. We say _XML-like_ because due to limitations of the macro system some concessions had to be made, see the examples below.
+The `xml!` macro accepts an XML-like syntax and transforms it into a `format_args!` invocation. We say _XML-like_ because due to limitations of declarative macros some concessions had to be made; see the examples below.
 
 Features of this crate include providing the value to be formatted inline in the formatting braces and control flow for conditionally formatting all in one simple package with zero dependencies!
 
-In your Cargo.toml add:
+In your Cargo.toml, add:
 
 ```
 [dependencies]
@@ -35,7 +35,7 @@ xml! {
 }.to_string()
 ```
 
-The resulting string is `<svg width="200" height="200"><line x1="0" y1="0" x2="20" y2="30" stroke="black" stroke-width="2" /><text x="30" y="20">Hello 'World!'</text></svg>`.
+The resulting string is `<svg width="200" height="200"><line x1="0" y1="0" x2="20" y2="30" stroke="black" stroke-width="2" /><text x="30" y="20">Hello 'World'!</text></svg>`.
 
 Note how the expression values to be formatted are inlined in the formatting braces.
 
@@ -51,7 +51,25 @@ xml! {
 
 The resulting string is `<span data-value="42">0x2a</span>`.
 
-Due to limitations of macros by example, a semicolon is used to separate the value from the formatting specifiers. The rules for the specifiers are exactly the same as the standard library of Rust.
+Due to limitations of declarative macros, a semicolon is used to separate the value from the formatting specifiers. The rules for the specifiers are exactly the same as [the standard library](https://doc.rust-lang.org/std/fmt/index.html) of Rust.
+
+### Composition
+
+```rust
+fn compose(f: &mut std::fmt::Formatter, a: i32) -> std::fmt::Result {
+	f.write_fmt(xml! {
+		<span>{a}</span>
+	})
+}
+
+xml! {
+	<p>|f| { compose(f, 42) }</p>
+}.to_string()
+```
+
+The resulting string is `<p><span>42</span></p>`.
+
+Closure syntax allows capturing of the underlying formatter like you would when writing a custom fmt trait. This enables to compose the final XML from reusable subtemplates.
 
 ### Supported tags
 
@@ -102,7 +120,7 @@ xml! {
 
 The resulting string is `<h1>Hello World</h1><b>13</b><ul><li>1*5=5</li><li>2*5=10</li><li>3*5=15</li><li>4*5=20</li><li>5*5=25</li></ul>`.
 
-Control flow are currently only supported outside tags. They are not supported in attributes. The expressions for `if` and `for` must be surrounded with parentheses due to macro by example limitations.
+Control flow are currently only supported outside tags. They are not supported in attributes. The expressions for `if` and `for` must be surrounded with parentheses due to declarative macro limitations.
 
 ### Specialised attribute syntax
 
@@ -130,7 +148,7 @@ For even more power closure syntax is available to write custom formatting code.
 Limitations
 -----------
 
-This crate is implemented with standard macros by example (`macro_rules!`). Because of this there are various limitations:
+This crate is implemented with declarative macros. Because of this, there are various limitations:
 
 * It is not possible to check whether tags are closed by the appropriate closing tag. This crate will happily accept `<open></close>`. It does enforce more simple lexical rules such as rejecting `</tag/>`.
 
@@ -141,6 +159,8 @@ This crate is implemented with standard macros by example (`macro_rules!`). Beca
 * The compiler may complain about macro expansion recursion limit being reached, simply apply the suggested fix and increase the limit. This crate implements a 'tt muncher' which are known to hit these limits.
 
 * Text nodes must be valid Rust literals. Bare words are not supported.
+
+* Braces must be escaped, eg. `"{{ }}"` results in a single set of `{ }`.
 
 License
 -------

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -40,6 +40,9 @@ impl<F: Fn(&mut fmt::Formatter) -> fmt::Result> fmt::Debug for FnFmt<F> {
 	}
 }
 
+// Prevent inlining the messy formatting code by moving it into a noinline function.
+// This could be done by applying the `#[inline(never)]` directly to the closure but this is experimental:
+// error[E0658]: attributes on expressions are experimental
 #[doc(hidden)]
 #[inline(never)]
 pub fn noinline<T, F: FnOnce() -> T>(f: F) -> T { f() }

--- a/src/template.rs
+++ b/src/template.rs
@@ -6,49 +6,49 @@ Replace the standard formatting macros using [template syntax](macro.template.ht
 #[doc(hidden)]
 #[macro_export]
 macro_rules! tprint {
-	($($tt:tt)+) => { $crate::noinline(|| print!("{}", $crate::FnFmt(|_f| { $crate::_template_!({_f} $($tt)*); Ok(()) }))) };
+	($($tt:tt)+) => { $crate::noinline(|| ::std::print!("{}", $crate::FnFmt(|_f| { $crate::_template_!({_f} $($tt)*); Ok(()) }))) };
 }
 /// Replaces `println!` using [template syntax](../macro.template.html).
 #[doc(hidden)]
 #[macro_export]
 macro_rules! tprintln {
-	($($tt:tt)+) => { $crate::noinline(|| println!("{}", $crate::FnFmt(|_f| { $crate::_template_!({_f} $($tt)*); Ok(()) }))) };
+	($($tt:tt)+) => { $crate::noinline(|| ::std::println!("{}", $crate::FnFmt(|_f| { $crate::_template_!({_f} $($tt)*); Ok(()) }))) };
 }
 /// Replaces `eprint!` using [template syntax](../macro.template.html).
 #[doc(hidden)]
 #[macro_export]
 macro_rules! teprint {
-	($($tt:tt)+) => { $crate::noinline(|| eprint!("{}", $crate::FnFmt(|_f| { $crate::_template_!({_f} $($tt)*); Ok(()) }))) };
+	($($tt:tt)+) => { $crate::noinline(|| ::std::eprint!("{}", $crate::FnFmt(|_f| { $crate::_template_!({_f} $($tt)*); Ok(()) }))) };
 }
 /// Replaces `eprintln!` using [template syntax](../macro.template.html).
 #[doc(hidden)]
 #[macro_export]
 macro_rules! teprintln {
-	($($tt:tt)+) => { $crate::noinline(|| eprintln!("{}", $crate::FnFmt(|_f| { $crate::_template_!({_f} $($tt)*); Ok(()) }))) };
+	($($tt:tt)+) => { $crate::noinline(|| ::std::eprintln!("{}", $crate::FnFmt(|_f| { $crate::_template_!({_f} $($tt)*); Ok(()) }))) };
 }
-/// Replaces `twrite!` using [template syntax](../macro.template.html).
+/// Replaces `write!` using [template syntax](../macro.template.html).
 #[doc(hidden)]
 #[macro_export]
 macro_rules! twrite {
-	($dst:expr, $($tt:tt)+) => { $crate::noinline(|| write!($dst, "{}", $crate::FnFmt(|_f| { $crate::_template_!({_f} $($tt)*); Ok(()) }))) };
+	($dst:expr, $($tt:tt)+) => { $crate::noinline(|| ::core::write!($dst, "{}", $crate::FnFmt(|_f| { $crate::_template_!({_f} $($tt)*); Ok(()) }))) };
 }
-/// Replaces `twriteln!` using [template syntax](../macro.template.html).
+/// Replaces `writeln!` using [template syntax](../macro.template.html).
 #[doc(hidden)]
 #[macro_export]
 macro_rules! twriteln {
-	($dst:expr, $($tt:tt)+) => { $crate::noinline(|| writeln!($dst, "{}", $crate::FnFmt(|_f| { $crate::_template_!({_f} $($tt)*); Ok(()) }))) };
+	($dst:expr, $($tt:tt)+) => { $crate::noinline(|| ::core::writeln!($dst, "{}", $crate::FnFmt(|_f| { $crate::_template_!({_f} $($tt)*); Ok(()) }))) };
 }
 /// Replaces `format!` using [template syntax](../macro.template.html).
 #[doc(hidden)]
 #[macro_export]
 macro_rules! tformat {
-	($($tt:tt)+) => { $crate::noinline(|| format!("{}", $crate::FnFmt(|_f| { $crate::_template_!({_f} $($tt)*); Ok(()) }))) };
+	($($tt:tt)+) => { $crate::noinline(|| ::std::format!("{}", $crate::FnFmt(|_f| { $crate::_template_!({_f} $($tt)*); Ok(()) }))) };
 }
 /// Replaces `panic!` using [template syntax](../macro.template.html).
 #[doc(hidden)]
 #[macro_export]
 macro_rules! tpanic {
-	($($tt:tt)+) => { $crate::noinline(|| panic!("{}", $crate::FnFmt(|_f| { $crate::_template_!({_f} $($tt)*); Ok(()) }))) };
+	($($tt:tt)+) => { $crate::noinline(|| ::core::panic!("{}", $crate::FnFmt(|_f| { $crate::_template_!({_f} $($tt)*); Ok(()) }))) };
 }
 
 pub use crate::{
@@ -61,3 +61,16 @@ pub use crate::{
 	tformat as format,
 	tpanic as panic,
 };
+
+#[test]
+fn test_noinline() {
+	use std::fmt::Write;
+	tprint!("print");
+	tprintln!("println");
+	teprint!("eprint");
+	teprintln!("eprintln");
+	let mut s = tformat!("format");
+	let _ = twrite!(s, "write");
+	let _ = twriteln!(s, "writeln");
+	// tpanic!("panic");
+}

--- a/src/template_impl.rs
+++ b/src/template_impl.rs
@@ -17,7 +17,7 @@
 /// The resulting string is `Hello World!`.
 ///
 /// Note how the expression values to be formatted are inlined in the formatting braces.
-/// Due to limitations of macros by example the formatting braces are not part of the surrounding string literals.
+/// Due to limitations of declarative macros the formatting braces are not part of the surrounding string literals.
 ///
 /// ### Formatting specifiers
 ///
@@ -31,7 +31,7 @@
 ///
 /// The resulting string is `hex(42) = 0x2a`.
 ///
-/// Due to limitations of macros by example, a semicolon is used to separate the value from the formatting specifiers.
+/// Due to limitations of declarative macros, a semicolon is used to separate the value from the formatting specifiers.
 /// The rules for the specifiers are exactly the same as Rust's standard formatting syntax.
 ///
 /// ### Control flow
@@ -86,7 +86,7 @@
 /// and the closure returns a [`std::fmt::Result`](https://doc.rust-lang.org/std/fmt/type.Result.html).
 #[macro_export]
 macro_rules! template {
-	($($tt:tt)*) => { format_args!("{}", $crate::FnFmt(|_f| { $crate::_template_!({_f} $($tt)*); Ok(()) })) };
+	($($tt:tt)*) => { ::core::format_args!("{}", $crate::FnFmt(|_f| { $crate::_template_!({_f} $($tt)*); Ok(()) })) };
 }
 
 #[macro_export]
@@ -101,10 +101,10 @@ macro_rules! _template_ {
 	};
 	// format
 	({$f:ident $($stmt:stmt)*} {$e:expr} $($tail:tt)*) => {
-		$crate::_template_!({$f $($stmt)* $f.write_fmt(format_args!("{}", $e))?} $($tail)*)
+		$crate::_template_!({$f $($stmt)* $f.write_fmt(::core::format_args!("{}", $e))?} $($tail)*)
 	};
 	({$f:ident $($stmt:stmt)*} {$e:expr;$($s:tt)*} $($tail:tt)*) => {
-		$crate::_template_!({$f $($stmt)* $f.write_fmt(format_args!(concat!("{:", $(stringify!($s),)* "}"), $e))?} $($tail)*)
+		$crate::_template_!({$f $($stmt)* $f.write_fmt(::core::format_args!(concat!("{:", $(stringify!($s),)* "}"), $e))?} $($tail)*)
 	};
 	({$f:ident $($stmt:stmt)*} |$ff:ident| { $($body:stmt)* } $($tail:tt)*) => {
 		$crate::_template_!({$f $($stmt)* { let $ff = &mut *$f; $($body)* }?} $($tail)*)

--- a/src/xml.rs
+++ b/src/xml.rs
@@ -1,54 +1,54 @@
 /*!
-Replace the standard formatting macros using [xml syntax](macro.xml.html).
+Replaces the standard formatting macros using [xml syntax](macro.xml.html).
 */
 
-/// Replaces `print!` using [xml syntax](macro.format_xml.html).
+/// Replaces `print!` using [xml syntax](macro.xml.html).
 #[doc(hidden)]
 #[macro_export]
 macro_rules! xprint {
-	($($tt:tt)+) => { $crate::noinline(|| print!("{}", $crate::format_xml!($($tt)*))) };
+	($($tt:tt)+) => { $crate::noinline(|| ::std::print!("{}", $crate::xml!($($tt)*))) };
 }
-/// Replaces `println!` using [xml syntax](macro.format_xml.html).
+/// Replaces `println!` using [xml syntax](macro.xml.html).
 #[doc(hidden)]
 #[macro_export]
 macro_rules! xprintln {
-	($($tt:tt)+) => { $crate::noinline(|| println!("{}", $crate::format_xml!($($tt)*))) };
+	($($tt:tt)+) => { $crate::noinline(|| ::std::println!("{}", $crate::xml!($($tt)*))) };
 }
-/// Replaces `eprint!` using [xml syntax](macro.format_xml.html).
+/// Replaces `eprint!` using [xml syntax](macro.xml.html).
 #[doc(hidden)]
 #[macro_export]
 macro_rules! xeprint {
-	($($tt:tt)+) => { $crate::noinline(|| eprint!("{}", $crate::format_xml!($($tt)*))) };
+	($($tt:tt)+) => { $crate::noinline(|| ::std::eprint!("{}", $crate::xml!($($tt)*))) };
 }
-/// Replaces `eprintln!` using [xml syntax](macro.format_xml.html).
+/// Replaces `eprintln!` using [xml syntax](macro.xml.html).
 #[doc(hidden)]
 #[macro_export]
 macro_rules! xeprintln {
-	($($tt:tt)+) => { $crate::noinline(|| eprintln!("{}", $crate::format_xml!($($tt)*))) };
+	($($tt:tt)+) => { $crate::noinline(|| ::std::eprintln!("{}", $crate::xml!($($tt)*))) };
 }
-/// Replaces `write!` using [xml syntax](macro.format_xml.html).
+/// Replaces `write!` using [xml syntax](macro.xml.html).
 #[doc(hidden)]
 #[macro_export]
 macro_rules! xwrite {
-	($dst:expr, $($tt:tt)+) => { $crate::noinline(|| write!($dst, "{}", $crate::format_xml!($($tt)*))) };
+	($dst:expr, $($tt:tt)+) => { $crate::noinline(|| ::core::write!($dst, "{}", $crate::xml!($($tt)*))) };
 }
-/// Replaces `writeln!` using [xml syntax](macro.format_xml.html).
+/// Replaces `writeln!` using [xml syntax](macro.xml.html).
 #[doc(hidden)]
 #[macro_export]
 macro_rules! xwriteln {
-	($dst:expr, $($tt:tt)+) => { $crate::noinline(|| writeln!($dst, "{}", $crate::format_xml!($($tt)*))) };
+	($dst:expr, $($tt:tt)+) => { $crate::noinline(|| ::core::writeln!($dst, "{}", $crate::xml!($($tt)*))) };
 }
-/// Replaces `format!` using [xml syntax](macro.format_xml.html).
+/// Replaces `format!` using [xml syntax](macro.xml.html).
 #[doc(hidden)]
 #[macro_export]
 macro_rules! xformat {
-	($($tt:tt)+) => { $crate::noinline(|| format!("{}", $crate::format_xml!($($tt)*))) };
+	($($tt:tt)+) => { $crate::noinline(|| ::std::format!("{}", $crate::xml!($($tt)*))) };
 }
-/// Replaces `panic!` using [xml syntax](macro.format_xml.html).
+/// Replaces `panic!` using [xml syntax](macro.xml.html).
 #[doc(hidden)]
 #[macro_export]
 macro_rules! xpanic {
-	($($tt:tt)+) => { $crate::noinline(|| panic!("{}", $crate::format_xml!($($tt)*))) };
+	($($tt:tt)+) => { $crate::noinline(|| ::core::panic!("{}", $crate::xml!($($tt)*))) };
 }
 
 pub use crate::{
@@ -61,3 +61,16 @@ pub use crate::{
 	xformat as format,
 	xpanic as panic,
 };
+
+#[test]
+fn test_noinline() {
+	use std::fmt::Write;
+	xprint!(<a>"print"</a>);
+	xprintln!(<a>"println"</a>);
+	xeprint!(<a>"eprint"</a>);
+	xeprintln!(<a>"eprintln"</a>);
+	let mut s = xformat!(<a>"format"</a>);
+	let _ = xwrite!(s, <a>"write"</a>);
+	let _ = xwriteln!(s, <a>"writeln"</a>);
+	// xpanic!(<a>"panic"</a>);
+}

--- a/src/xml_impl.rs
+++ b/src/xml_impl.rs
@@ -1,4 +1,4 @@
-/// Template XML.
+/// Template XML-like.
 ///
 /// Compiles down to a single `format_args!` invocation.
 ///
@@ -47,7 +47,29 @@
 ///
 /// The resulting string is `<span data-value="42">0x2a</span>`.
 ///
-/// Due to limitations of macros by example, a semicolon is used to separate the value from the formatting specifiers. The rules for the specifiers are exactly the same as the standard library of Rust.
+/// Due to limitations of declarative macros, a semicolon is used to separate the value from the formatting specifiers. The rules for the specifiers are exactly the same as [the standard library](https://doc.rust-lang.org/std/fmt/index.html) of Rust.
+///
+/// ### Composition
+///
+/// ```rust
+/// # use format_xml::xml;
+/// #
+/// fn compose(f: &mut std::fmt::Formatter, a: i32) -> std::fmt::Result {
+/// 	f.write_fmt(xml! {
+/// 		<span>{a}</span>
+/// 	})
+/// }
+///
+/// # let result =
+/// xml! {
+/// 	<p>|f| { compose(f, 42) }</p>
+/// }.to_string()
+/// # ; assert_eq!(result, r#"<p><span>42</span></p>"#);
+/// ```
+///
+/// The resulting string is `<p><span>42</span></p>`.
+///
+/// Closure syntax allows capturing of the underlying formatter like you would when writing a custom fmt trait. This enables to compose the final XML from reusable subtemplates.
 ///
 /// ### Supported tags
 ///
@@ -106,7 +128,7 @@
 ///
 /// The resulting string is `<h1>Hello World</h1><b>13</b><ul><li>1*5=5</li><li>2*5=10</li><li>3*5=15</li><li>4*5=20</li><li>5*5=25</li></ul>`.
 ///
-/// Control flow are currently only supported outside tags. They are not supported in attributes. The expressions for `if` and `for` must be surrounded with parentheses due to macro by example limitations.
+/// Control flow are currently only supported outside tags. They are not supported in attributes. The expressions for `if` and `for` must be surrounded with parentheses due to declarative macro limitations.
 ///
 /// ### Specialised attribute syntax
 ///
@@ -138,7 +160,7 @@
 /// Limitations
 /// -----------
 ///
-/// This crate is implemented with standard macros by example (`macro_rules!`). Because of this there are various limitations:
+/// This crate is implemented with declarative macros. Because of this, there are various limitations:
 ///
 /// * It is not possible to check whether tags are closed by the appropriate closing tag. This crate will happily accept `<open></close>`. It does enforce more simple lexical rules such as rejecting `</tag/>`.
 ///
@@ -150,10 +172,11 @@
 ///
 /// * Text nodes must be valid Rust literals. Bare words are not supported.
 ///
+/// * Braces must be escaped, eg. `"{{ }}"` results in a single set of `{ }`.
 #[macro_export]
 macro_rules! xml {
 	($($tt:tt)*) => {
-		$crate::_format_tag1_!(; "",; $($tt)*)
+		$crate::_format_tag1_!(; concat!(),; $($tt)*)
 	};
 }
 
@@ -164,80 +187,80 @@ macro_rules! xml {
 #[macro_export]
 macro_rules! _format_tag1_ {
 	// tags
-	(; $fmt:expr, $($args:expr,)*; </ $($tail:tt)*) => {
-		$crate::_format_ident1_!(_format_tag3_!; concat!($fmt, "</"), $($args,)*; $($tail)*)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*; </ $($tail:tt)*) => {
+		$crate::_format_ident1_!(_format_tag3_!; concat!($($fmt,)* "</",), $($args,)*; $($tail)*)
 	};
-	(; $fmt:expr, $($args:expr,)*; <? $($tail:tt)*) => {
-		$crate::_format_ident1_!(_format_attrs1_! _format_tag4_!; concat!($fmt, "<?"), $($args,)*; $($tail)*)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*; <? $($tail:tt)*) => {
+		$crate::_format_ident1_!(_format_attrs1_! _format_tag4_!; concat!($($fmt,)* "<?",), $($args,)*; $($tail)*)
 	};
-	(; $fmt:expr, $($args:expr,)*; <!-- $($tail:tt)*) => {
-		$crate::_format_text1_!(; concat!($fmt, "<!-- "), $($args,)*; $($tail)*)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*; <!-- $($tail:tt)*) => {
+		$crate::_format_text1_!(; concat!($($fmt,)* "<!-- ",), $($args,)*; $($tail)*)
 	};
-	(; $fmt:expr, $($args:expr,)*; <![CDATA[ $($text:literal)* ]]> $($tail:tt)*) => {
-		$crate::_format_tag1_!(; concat!($fmt, "<![CDATA[", $($text,)* "]]>"), $($args,)*; $($tail)*)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*; <![CDATA[ $($text:literal)* ]]> $($tail:tt)*) => {
+		$crate::_format_tag1_!(; concat!($($fmt,)* "<![CDATA[", $($text,)* "]]>",), $($args,)*; $($tail)*)
 	};
-	(; $fmt:expr, $($args:expr,)*; <! $($tail:tt)*) => {
-		$crate::_format_ident1_!(_format_attrs1_! _format_tag2_!; concat!($fmt, "<!"), $($args,)*; $($tail)*)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*; <! $($tail:tt)*) => {
+		$crate::_format_ident1_!(_format_attrs1_! _format_tag2_!; concat!($($fmt,)* "<!",), $($args,)*; $($tail)*)
 	};
-	(; $fmt:expr, $($args:expr,)*; < $($tail:tt)*) => {
-		$crate::_format_ident1_!(_format_attrs1_! _format_tag2_!; concat!($fmt, "<"), $($args,)*; $($tail)*)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*; < $($tail:tt)*) => {
+		$crate::_format_ident1_!(_format_attrs1_! _format_tag2_!; concat!($($fmt,)* "<",), $($args,)*; $($tail)*)
 	};
 	// text
-	(; $fmt:expr, $($args:expr,)*; $text:literal $($tail:tt)*) => {
-		$crate::_format_tag1_!(; concat!($fmt, $text), $($args,)*; $($tail)*)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*; $text:literal $($tail:tt)*) => {
+		$crate::_format_tag1_!(; concat!($($fmt,)* $text,), $($args,)*; $($tail)*)
 	};
-	(; $fmt:expr, $($args:expr,)*; {$e:expr} $($tail:tt)*) => {
-		$crate::_format_tag1_!(; concat!($fmt, "{}"), $($args,)* $e,; $($tail)*)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*; {$e:expr} $($tail:tt)*) => {
+		$crate::_format_tag1_!(; concat!($($fmt,)* "{}",), $($args,)* $e,; $($tail)*)
 	};
-	(; $fmt:expr, $($args:expr,)*; {$e:expr;$($s:tt)*} $($tail:tt)*) => {
-		$crate::_format_tag1_!(; concat!($fmt, "{:", $(stringify!($s),)* "}"), $($args,)* $e,; $($tail)*)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*; {$e:expr;$($s:tt)*} $($tail:tt)*) => {
+		$crate::_format_tag1_!(; concat!($($fmt,)* "{:", $(stringify!($s),)* "}",), $($args,)* $e,; $($tail)*)
 	};
-	(; $fmt:expr, $($args:expr,)*; escape!($($body:tt)*) $($tail:tt)*) => {
-		$crate::_format_tag1_!(; concat!($fmt, "{}"), $($args,)* $crate::escape!($($body)*),; $($tail)*)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*; escape!($($body:tt)*) $($tail:tt)*) => {
+		$crate::_format_tag1_!(; concat!($($fmt,)* "{}",), $($args,)* $crate::escape!($($body)*),; $($tail)*)
 	};
-	(; $fmt:expr, $($args:expr,)*; |$f:ident| { $($body:tt)* } $($tail:tt)*) => {
-		$crate::_format_tag1_!(; concat!($fmt, "{}"), $($args,)* $crate::FnFmt(|$f| { $($body)* }),; $($tail)*)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*; |$f:ident| { $($body:tt)* } $($tail:tt)*) => {
+		$crate::_format_tag1_!(; concat!($($fmt,)* "{}",), $($args,)* $crate::FnFmt(|$f| { $($body)* }),; $($tail)*)
 	};
 	// control
-	(; $fmt:expr, $($args:expr,)*; let $p:pat = $e:expr; $($tail:tt)*) => {
-		$crate::_format_tag1_!(; concat!($fmt, "{}"), $($args,)* $crate::FnFmt(|f| match $e { $p => f.write_fmt($crate::xml!{$($tail)*}) }),;)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*; let $p:pat = $e:expr; $($tail:tt)*) => {
+		$crate::_format_tag1_!(; concat!($($fmt,)* "{}",), $($args,)* $crate::FnFmt(|f| match $e { $p => f.write_fmt($crate::xml!{$($tail)*}) }),;)
 	};
-	(; $fmt:expr, $($args:expr,)*; if $($tail:tt)*) => {
-		$crate::_format_if1_!(; $fmt, $($args,)*; f []; if $($tail)*)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*; if $($tail:tt)*) => {
+		$crate::_format_if1_!(; concat!($($fmt,)*), $($args,)*; f []; if $($tail)*)
 	};
-	(; $fmt:expr, $($args:expr,)*; match ($e:expr) { $($p:pat => { $($body:tt)* })* } $($tail:tt)*) => {
-		$crate::_format_tag1_!(; concat!($fmt, "{}"), $($args,)* $crate::FnFmt(|f| match $e { $($p => f.write_fmt($crate::xml!{$($body)*}),)* }),; $($tail)*)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*; match ($e:expr) { $($p:pat => { $($body:tt)* })* } $($tail:tt)*) => {
+		$crate::_format_tag1_!(; concat!($($fmt,)* "{}",), $($args,)* $crate::FnFmt(|f| match $e { $($p => f.write_fmt($crate::xml!{$($body)*}),)* }),; $($tail)*)
 	};
-	(; $fmt:expr, $($args:expr,)*; for $p:pat in ($e:expr) { $($body:tt)* } $($tail:tt)*) => {
-		$crate::_format_tag1_!(; concat!($fmt, "{}"), $($args,)* $crate::FnFmt(|f| { for $p in $e { f.write_fmt($crate::xml!{$($body)*})?; } Ok(()) }),; $($tail)*)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*; for $p:pat in ($e:expr) { $($body:tt)* } $($tail:tt)*) => {
+		$crate::_format_tag1_!(; concat!($($fmt,)* "{}",), $($args,)* $crate::FnFmt(|f| { for $p in $e { f.write_fmt($crate::xml!{$($body)*})?; } Ok(()) }),; $($tail)*)
 	};
 	// term
-	(; $fmt:expr, $($args:expr,)*;) => {
-		format_args!($fmt $(,$args)*)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*;) => {
+		::core::format_args!(concat!($($fmt,)*) $(,$args)*)
 	};
 }
 #[doc(hidden)]
 #[macro_export]
 macro_rules! _format_tag2_ {
-	(; $fmt:expr, $($args:expr,)*; > $($tail:tt)*) => {
-		$crate::_format_tag1_!(; concat!($fmt, ">"), $($args,)*; $($tail)*)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*; > $($tail:tt)*) => {
+		$crate::_format_tag1_!(; concat!($($fmt,)* ">",), $($args,)*; $($tail)*)
 	};
-	(; $fmt:expr, $($args:expr,)*; /> $($tail:tt)*) => {
-		$crate::_format_tag1_!(; concat!($fmt, " />"), $($args,)*; $($tail)*)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*; /> $($tail:tt)*) => {
+		$crate::_format_tag1_!(; concat!($($fmt,)* " />",), $($args,)*; $($tail)*)
 	};
 }
 #[doc(hidden)]
 #[macro_export]
 macro_rules! _format_tag3_ {
-	(; $fmt:expr, $($args:expr,)*; > $($tail:tt)*) => {
-		$crate::_format_tag1_!(; concat!($fmt, ">"), $($args,)*; $($tail)*)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*; > $($tail:tt)*) => {
+		$crate::_format_tag1_!(; concat!($($fmt,)* ">",), $($args,)*; $($tail)*)
 	};
 }
 #[doc(hidden)]
 #[macro_export]
 macro_rules! _format_tag4_ {
-	(; $fmt:expr, $($args:expr,)*; ?> $($tail:tt)*) => {
-		$crate::_format_tag1_!(; concat!($fmt, "?>"), $($args,)*; $($tail)*)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*; ?> $($tail:tt)*) => {
+		$crate::_format_tag1_!(; concat!($($fmt,)* "?>",), $($args,)*; $($tail)*)
 	};
 }
 
@@ -247,27 +270,27 @@ macro_rules! _format_tag4_ {
 #[doc(hidden)]
 #[macro_export]
 macro_rules! _format_if1_ {
-	(; $fmt:expr, $($args:expr,)*; $f:ident [$($c:tt)*]; if ($e:expr) { $($body:tt)* } $($tail:tt)*) => {
-		$crate::_format_if2_!(; $fmt, $($args,)*; $f [$($c)* [if ($e) { $f.write_fmt($crate::xml!{$($body)*}) }]]; $($tail)*)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*; $f:ident [$($c:tt)*]; if ($e:expr) { $($body:tt)* } $($tail:tt)*) => {
+		$crate::_format_if2_!(; concat!($($fmt,)*), $($args,)*; $f [$($c)* [if ($e) { $f.write_fmt($crate::xml!{$($body)*}) }]]; $($tail)*)
 	};
-	(; $fmt:expr, $($args:expr,)*; $f:ident [$($c:tt)*]; if let $p:pat = ($e:expr) { $($body:tt)* } $($tail:tt)*) => {
-		$crate::_format_if2_!(; $fmt, $($args,)*; $f [$($c)* [if let $p = ($e) { $f.write_fmt($crate::xml!{$($body)*}) }]]; $($tail)*)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*; $f:ident [$($c:tt)*]; if let $p:pat = ($e:expr) { $($body:tt)* } $($tail:tt)*) => {
+		$crate::_format_if2_!(; concat!($($fmt,)*), $($args,)*; $f [$($c)* [if let $p = ($e) { $f.write_fmt($crate::xml!{$($body)*}) }]]; $($tail)*)
 	};
 }
 #[doc(hidden)]
 #[macro_export]
 macro_rules! _format_if2_ {
-	(; $fmt:expr, $($args:expr,)*; $f:ident [$($c:tt)*]; else if ($e:expr) { $($body:tt)* } $($tail:tt)*) => {
-		$crate::_format_if2_!(; $fmt, $($args,)*; $f [$($c)* [else if ($e) { $f.write_fmt($crate::xml!{$($body)*}) }]]; $($tail)*)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*; $f:ident [$($c:tt)*]; else if ($e:expr) { $($body:tt)* } $($tail:tt)*) => {
+		$crate::_format_if2_!(; concat!($($fmt,)*), $($args,)*; $f [$($c)* [else if ($e) { $f.write_fmt($crate::xml!{$($body)*}) }]]; $($tail)*)
 	};
-	(; $fmt:expr, $($args:expr,)*; $f:ident [$($c:tt)*]; else if let $p:pat = ($e:expr) { $($body:tt)* } $($tail:tt)*) => {
-		$crate::_format_if2_!(; $fmt, $($args,)*; $f [$($c)* [else if let $p = ($e) { $f.write_fmt($crate::xml!{$($body)*}) }]]; $($tail)*)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*; $f:ident [$($c:tt)*]; else if let $p:pat = ($e:expr) { $($body:tt)* } $($tail:tt)*) => {
+		$crate::_format_if2_!(; concat!($($fmt,)*), $($args,)*; $f [$($c)* [else if let $p = ($e) { $f.write_fmt($crate::xml!{$($body)*}) }]]; $($tail)*)
 	};
-	(; $fmt:expr, $($args:expr,)*; $f:ident [$([$($c:tt)*])*]; else { $($body:tt)* } $($tail:tt)*) => {
-		$crate::_format_tag1_!(; concat!($fmt, "{}"), $($args,)* $crate::FnFmt(|$f| $($($c)*)* else { $f.write_fmt($crate::xml!{$($body)*}) }),; $($tail)*)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*; $f:ident [$([$($c:tt)*])*]; else { $($body:tt)* } $($tail:tt)*) => {
+		$crate::_format_tag1_!(; concat!($($fmt,)* "{}",), $($args,)* $crate::FnFmt(|$f| $($($c)*)* else { $f.write_fmt($crate::xml!{$($body)*}) }),; $($tail)*)
 	};
-	(; $fmt:expr, $($args:expr,)*; $f:ident [$([$($c:tt)*])*]; $($tail:tt)*) => {
-		$crate::_format_tag1_!(; concat!($fmt, "{}"), $($args,)* $crate::FnFmt(|$f| $($($c)*)* else { Ok(()) }),; $($tail)*)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*; $f:ident [$([$($c:tt)*])*]; $($tail:tt)*) => {
+		$crate::_format_tag1_!(; concat!($($fmt,)* "{}",), $($args,)* $crate::FnFmt(|$f| $($($c)*)* else { Ok(()) }),; $($tail)*)
 	};
 }
 
@@ -277,17 +300,17 @@ macro_rules! _format_if2_ {
 #[doc(hidden)]
 #[macro_export]
 macro_rules! _format_text1_ {
-	(; $fmt:expr, $($args:expr,)*; --> $($tail:tt)*) => {
-		$crate::_format_tag1_!(; concat!($fmt, " -->"), $($args,)*; $($tail)*)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*; --> $($tail:tt)*) => {
+		$crate::_format_tag1_!(; concat!($($fmt,)* " -->",), $($args,)*; $($tail)*)
 	};
-	(; $fmt:expr, $($args:expr,)*; $text:literal $($tail:tt)*) => {
-		$crate::_format_text1_!(; concat!($fmt, $text), $($args,)*; $($tail)*)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*; $text:literal $($tail:tt)*) => {
+		$crate::_format_text1_!(; concat!($($fmt,)* $text,), $($args,)*; $($tail)*)
 	};
-	(; $fmt:expr, $($args:expr,)*; {$e:expr} $($tail:tt)*) => {
-		$crate::_format_text1_!(; concat!($fmt, "{}"), $($args,)* $e,; $($tail)*)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*; {$e:expr} $($tail:tt)*) => {
+		$crate::_format_text1_!(; concat!($($fmt,)* "{}",), $($args,)* $e,; $($tail)*)
 	};
-	(; $fmt:expr, $($args:expr,)*; {$e:expr;$($s:tt)*} $($tail:tt)*) => {
-		$crate::_format_text1_!(; concat!($fmt, "{:", $(stringify!($s),)* "}"), $($args,)* $e,; $($tail)*)
+	(; concat!($($fmt:expr,)*), $($args:expr,)*; {$e:expr;$($s:tt)*} $($tail:tt)*) => {
+		$crate::_format_text1_!(; concat!($($fmt,)* "{:", $(stringify!($s),)* "}",), $($args,)* $e,; $($tail)*)
 	};
 }
 
@@ -297,24 +320,24 @@ macro_rules! _format_text1_ {
 #[doc(hidden)]
 #[macro_export]
 macro_rules! _format_ident1_ {
-	($($q:ident!)+; $fmt:expr, $($args:expr,)*; $name:ident : $($tail:tt)*) => {
-		$crate::_format_ident2_!($($q!)+; concat!($fmt, stringify!($name), ":"), $($args,)*; $($tail)*)
+	($($q:ident!)+; concat!($($fmt:expr,)*), $($args:expr,)*; $name:ident : $($tail:tt)*) => {
+		$crate::_format_ident2_!($($q!)+; concat!($($fmt,)* stringify!($name), ":",), $($args,)*; $($tail)*)
 	};
-	($($q:ident!)+; $fmt:expr, $($args:expr,)*; $name:ident - $($tail:tt)*) => {
-		$crate::_format_ident2_!($($q!)+; concat!($fmt, stringify!($name), "-"), $($args,)*; $($tail)*)
+	($($q:ident!)+; concat!($($fmt:expr,)*), $($args:expr,)*; $name:ident - $($tail:tt)*) => {
+		$crate::_format_ident2_!($($q!)+; concat!($($fmt,)* stringify!($name), "-",), $($args,)*; $($tail)*)
 	};
-	($next:ident! $($q:ident!)*; $fmt:expr, $($args:expr,)*; $name:ident $($tail:tt)*) => {
-		$crate::$next!($($q!)*; concat!($fmt, stringify!($name)), $($args,)*; $($tail)*)
+	($next:ident! $($q:ident!)*; concat!($($fmt:expr,)*), $($args:expr,)*; $name:ident $($tail:tt)*) => {
+		$crate::$next!($($q!)*; concat!($($fmt,)* stringify!($name),), $($args,)*; $($tail)*)
 	};
 }
 #[doc(hidden)]
 #[macro_export]
 macro_rules! _format_ident2_ {
-	($($q:ident!)+; $fmt:expr, $($args:expr,)*; $name:ident - $($tail:tt)*) => {
-		$crate::_format_ident2_!($($q!)+; concat!($fmt, stringify!($name), "-"), $($args,)*; $($tail)*)
+	($($q:ident!)+; concat!($($fmt:expr,)*), $($args:expr,)*; $name:ident - $($tail:tt)*) => {
+		$crate::_format_ident2_!($($q!)+; concat!($($fmt,)* stringify!($name), "-",), $($args,)*; $($tail)*)
 	};
-	($next:ident! $($q:ident!)*; $fmt:expr, $($args:expr,)*; $name:ident $($tail:tt)*) => {
-		$crate::$next!($($q!)*; concat!($fmt, stringify!($name)), $($args,)*; $($tail)*)
+	($next:ident! $($q:ident!)*; concat!($($fmt:expr,)*), $($args:expr,)*; $name:ident $($tail:tt)*) => {
+		$crate::$next!($($q!)*; concat!($($fmt,)* stringify!($name),), $($args,)*; $($tail)*)
 	};
 }
 
@@ -324,38 +347,38 @@ macro_rules! _format_ident2_ {
 #[doc(hidden)]
 #[macro_export]
 macro_rules! _format_attrs1_ {
-	($($q:ident!)+; $fmt:expr, $($args:expr,)*; $tail_id:ident $($tail:tt)*) => {
-		$crate::_format_ident1_!(_format_attrs2_! $($q!)+; concat!($fmt, " "), $($args,)*; $tail_id $($tail)*)
+	($($q:ident!)+; concat!($($fmt:expr,)*), $($args:expr,)*; $tail_id:ident $($tail:tt)*) => {
+		$crate::_format_ident1_!(_format_attrs2_! $($q!)+; concat!($($fmt,)* " ",), $($args,)*; $tail_id $($tail)*)
 	};
-	($next:ident! $($q:ident!)*; $fmt:expr, $($args:expr,)*; $($tail:tt)*) => {
-		$crate::$next!($($q!)*; $fmt, $($args,)*; $($tail)*)
+	($next:ident! $($q:ident!)*; concat!($($fmt:expr,)*), $($args:expr,)*; $($tail:tt)*) => {
+		$crate::$next!($($q!)*; concat!($($fmt,)*), $($args,)*; $($tail)*)
 	};
 }
 #[doc(hidden)]
 #[macro_export]
 macro_rules! _format_attrs2_ {
-	($($q:ident!)+; $fmt:expr, $($args:expr,)*; = $text:literal $($tail:tt)*) => {
-		$crate::_format_attrs1_!($($q!)*; concat!($fmt, "=\"", $text, '"'), $($args,)*; $($tail)*)
+	($($q:ident!)+; concat!($($fmt:expr,)*), $($args:expr,)*; = $text:literal $($tail:tt)*) => {
+		$crate::_format_attrs1_!($($q!)*; concat!($($fmt,)* "=\"", $text, '"',), $($args,)*; $($tail)*)
 	};
-	($($q:ident!)+; $fmt:expr, $($args:expr,)*; = {$e:expr} $($tail:tt)*) => {
-		$crate::_format_attrs1_!($($q!)*; concat!($fmt, "=\"{}\""), $($args,)* $e,; $($tail)*)
+	($($q:ident!)+; concat!($($fmt:expr,)*), $($args:expr,)*; = {$e:expr} $($tail:tt)*) => {
+		$crate::_format_attrs1_!($($q!)*; concat!($($fmt,)* "=\"{}\"",), $($args,)* $e,; $($tail)*)
 	};
-	($($q:ident!)+; $fmt:expr, $($args:expr,)*; = {$e:expr; $($s:tt)*} $($tail:tt)*) => {
-		$crate::_format_attrs1_!($($q!)*; concat!($fmt, "=\"{:", $(stringify!($s),)* "}\""), $($args,)* $e,; $($tail)*)
+	($($q:ident!)+; concat!($($fmt:expr,)*), $($args:expr,)*; = {$e:expr; $($s:tt)*} $($tail:tt)*) => {
+		$crate::_format_attrs1_!($($q!)*; concat!($($fmt,)* "=\"{:", $(stringify!($s),)* "}\"",), $($args,)* $e,; $($tail)*)
 	};
-	($($q:ident!)+; $fmt:expr, $($args:expr,)*; = [$($text:literal : $cond:expr),*$(,)?] $($tail:tt)*) => {
-		$crate::_format_attrs1_!($($q!)*; concat!($fmt, "=\"{}\""), $($args,)* $crate::FnFmt(|f| { $(if $cond { f.write_str(concat!($text, " "))? })* Ok(()) }),; $($tail)*)
+	($($q:ident!)+; concat!($($fmt:expr,)*), $($args:expr,)*; = [$($text:literal : $cond:expr),*$(,)?] $($tail:tt)*) => {
+		$crate::_format_attrs1_!($($q!)*; concat!($($fmt,)* "=\"{}\"",), $($args,)* $crate::FnFmt(|f| { $(if $cond { f.write_str(concat!($text, " "))? })* Ok(()) }),; $($tail)*)
 	};
-	($($q:ident!)+; $fmt:expr, $($args:expr,)*; = ($($body:tt)*) $($tail:tt)*) => {
-		$crate::_format_attrs1_!($($q!)*; concat!($fmt, "=\"{}\""), $($args,)* $crate::FnFmt(|f| f.write_fmt($crate::template!($($body)*))),; $($tail)*)
+	($($q:ident!)+; concat!($($fmt:expr,)*), $($args:expr,)*; = ($($body:tt)*) $($tail:tt)*) => {
+		$crate::_format_attrs1_!($($q!)*; concat!($($fmt,)* "=\"{}\"",), $($args,)* $crate::FnFmt(|f| f.write_fmt($crate::template!($($body)*))),; $($tail)*)
 	};
-	($($q:ident!)+; $fmt:expr, $($args:expr,)*; = escape!($($body:tt)*) $($tail:tt)*) => {
-		$crate::_format_attrs1_!($($q!)*; concat!($fmt, "=\"{}\""), $($args,)* $crate::escape!($($body)*),; $($tail)*)
+	($($q:ident!)+; concat!($($fmt:expr,)*), $($args:expr,)*; = escape!($($body:tt)*) $($tail:tt)*) => {
+		$crate::_format_attrs1_!($($q!)*; concat!($($fmt,)* "=\"{}\"",), $($args,)* $crate::escape!($($body)*),; $($tail)*)
 	};
-	($($q:ident!)+; $fmt:expr, $($args:expr,)*; = |$f:ident| { $($body:tt)* } $($tail:tt)*) => {
-		$crate::_format_attrs1_!($($q!)*; concat!($fmt, "=\"{}\""), $($args,)* $crate::FnFmt(|$f| { $($body)* }),; $($tail)*)
+	($($q:ident!)+; concat!($($fmt:expr,)*), $($args:expr,)*; = |$f:ident| { $($body:tt)* } $($tail:tt)*) => {
+		$crate::_format_attrs1_!($($q!)*; concat!($($fmt,)* "=\"{}\"",), $($args,)* $crate::FnFmt(|$f| { $($body)* }),; $($tail)*)
 	};
-	($($q:ident!)+; $fmt:expr, $($args:expr,)*; $($tail:tt)*) => {
-		$crate::_format_attrs1_!($($q!)*; $fmt, $($args,)*; $($tail)*)
+	($($q:ident!)+; concat!($($fmt:expr,)*), $($args:expr,)*; $($tail:tt)*) => {
+		$crate::_format_attrs1_!($($q!)*; concat!($($fmt,)*), $($args,)*; $($tail)*)
 	};
 }

